src/mil_common/electrical_protocol_cpp/include/electrical_protocol_cpp/packet.h

```diff
@@ -0,0 +1,188 @@
+#include <exception>
+
+#include <cppystruct.h>
+// #include <cppystruct/format.h>
+
+namespace electrical_protocol
+{
+    class SerialDevice;
+
+    class Packet
+    {
+        public:
+        friend class SerialDevice;
+        Packet() = delete;
+        ~Packet() noexcept
+        {
+
+        }
+        
+        Packet(uint8_t classId, uint8_t subClassId) noexcept :id_({classId, subClassId})
+        {
+        }
+
+        explicit Packet(const std::pair<uint8_t, uint8_t>& id) noexcept :id_(id) 
+        {
+        }
+
+        Packet(Packet&& packet) noexcept : id_(packet.id_)
+        {
+            data_ = std::move(packet.data_);
+        }
+
+        Packet(const Packet& packet) noexcept : id_(packet.id_)
+        {
+            data_ = packet.data_;
+        }
+
+        inline Packet& operator= (Packet&& packet)
+        {
+            if(id_ != packet.id_)
+                throw std::runtime_error("= between two packets with different id is impossiable");
+            
+            data_ = std::move(packet.data_);
+            return *this;
+        }
+
+        inline Packet& operator= (Packet& packet)
+        {
+            if(id_ != packet.id_)
+                throw std::runtime_error("= between two packets with different id is impossiable");
+            
+            data_ = packet.data_;
+            
+            return *this;
+        }
+
+        inline uint8_t& operator[] (size_t index)
+        {
+            if(index + HEADER_LEN + TRAILER_LEN >= data_.size())
+                throw std::out_of_range("No enough data");
+
+            return data_[index + HEADER_LEN];
+        }
+
+        inline const uint8_t& operator[] (size_t index) const
+        {
+            if(index + HEADER_LEN + TRAILER_LEN >= data_.size())
+                throw std::out_of_range("No enough data");
+
+            return data_[index + HEADER_LEN];
+        }
+
+        struct IdHash
+        {
+            size_t operator()(const std::pair<uint8_t, uint8_t>& p) const
+            {
+                return static_cast<size_t>((static_cast<uint16_t>(p.first) << 8) + p.second);
+            }
+        };
+
+        inline const std::pair<uint8_t, uint8_t>& id() const noexcept
+        {
+            return id_;
+        }
+
+        inline size_t size() const noexcept
+        {
+            if(data_.size() < (HEADER_LEN + TRAILER_LEN))
+                return 0;
+
+            return data_.size() - HEADER_LEN - TRAILER_LEN;
+        }
+
+        inline void resize(size_t size) noexcept
+        {
+            size_t orignalSize = data_.size();
+            data_.resize(size + HEADER_LEN + TRAILER_LEN);
+            if(orignalSize == 0)
+            {
+                data_[0] = SYNC_CHAR_1;
+                data_[1] = SYNC_CHAR_2;
+                data_[2] = id_.first;
+                data_[3] = id_.second;
+            }
+        }
+        
+        template <typename Fmt, typename... Args>
+        void pack(Fmt, Args&&... args)
```

Comment:
If you use the CRTP method suggested above, you could consider calling a method like `get_payload` from the derived class, and then use this in the packed format of the packet.